framework/src/transfers/MultiAppMeshFunctionTransfer.C

```diff
@@ -63,289 +62,347 @@ MultiAppMeshFunctionTransfer::execute()
+          {
+            if (bboxes[i_from].contains_point(centroid + _to_positions[i_to]))
             {
-              Elem * elem = *elem_it;
-
-              Point centroid = elem->centroid();
-
-              if (elem->n_dofs(sys_num, var_num) > 0) // If this variable has dofs at this elem
-              {
-                // The zero only works for LAGRANGE!
-                dof_id_type dof = elem->dof_number(sys_num, var_num, 0);
-
-                // Swap back
-                Moose::swapLibMeshComm(swapped);
-                Real from_value = from_func(centroid+_multi_app->position(i));
-                // Swap again
-                swapped = Moose::swapLibMeshComm(_multi_app->comm());
-
-                if (from_value != OutOfMeshValue)
-                  solution.set(dof, from_value);
-                else if (_error_on_miss)
-                  mooseError("Point not found! " << centroid+_multi_app->position(i) << std::endl);
-              }
+              std::pair<unsigned int, unsigned int> key(i_to, elem->id());
+              point_index_map[i_proc][key] = outgoing_points[i_proc].size();
+              outgoing_points[i_proc].push_back(centroid + _to_positions[i_to]);
+              point_found = true;
             }
           }
-
-          solution.close();
-          to_sys->update();
-
-          // Swap back
-          Moose::swapLibMeshComm(swapped);
         }
       }
+    }
+  }
 
-      delete serialized_solution;
+  ////////////////////
+  // Request point evaluations from other processors and handle requests sent to
+  // this processor.
+  ////////////////////
 
-      break;
-    }
-    case FROM_MULTIAPP:
+  // Get the local bounding boxes.
+  std::vector<MeshTools::BoundingBox> local_bboxes(froms_per_proc[processor_id()]);
+  {
+    // Find the index to the first of this processor's local bounding boxes.
+    unsigned int local_start = 0;
+    for (processor_id_type i_proc = 0;
+         i_proc < n_processors() && i_proc != processor_id();
+         i_proc++)
     {
-      // TODO: This doesn't work with the master app being displaced see #3424
-      if (_displaced_target_mesh)
-        mooseError("displaced_target_mesh is not yet implemented for transferring 'from_multiapp'");
-
-      FEProblem & to_problem = *_multi_app->problem();
-      MooseVariable & to_var = to_problem.getVariable(0, _to_var_name);
-      SystemBase & to_system_base = to_var.sys();
+      local_start += froms_per_proc[i_proc];
+    }
 
-      System & to_sys = to_system_base.system();
+    // Extract the local bounding boxes.
+    for (unsigned int i_from = 0; i_from < froms_per_proc[processor_id()]; i_from++)
+    {
+      local_bboxes[i_from] = bboxes[local_start + i_from];
+    }
+  }
 
-      unsigned int to_sys_num = to_sys.number();
+  // Setup the local mesh functions.
+  std::vector<MeshFunction *> local_meshfuns(_from_problems.size(), NULL);
```

Comment:
Consider using `MooseSharedPointer<MeshFunction>` instead of raw pointers in this vector.  That way you don't have to worry about deleting them later.